assembl/static2/js/app/reducers/adminReducer/voteSession.js

```diff
@@ -325,22 +372,119 @@ export const tokenCategoriesById = (state: Map<string, Map> = Map(), action: Red
   }
 };
 
+export const voteProposalsHaveChanged = (state: boolean = false, action: ReduxAction<Action>) => {
+  switch (action.type) {
+  case CREATE_VOTE_PROPOSAL:
+  case DELETE_VOTE_PROPOSAL:
+  case UPDATE_VOTE_PROPOSAL_TITLE:
+  case UPDATE_VOTE_PROPOSAL_DESCRIPTION:
+  case ADD_MODULE_TO_PROPOSAL:
+  case DELETE_VOTE_MODULE:
+    return true;
+  case UPDATE_VOTE_PROPOSALS:
+    return false;
+  default:
+    return state;
+  }
+};
+
+export const voteProposalsInOrder = (state: List<number> = List(), action: ReduxAction<Action>) => {
+  switch (action.type) {
+  case UPDATE_VOTE_PROPOSALS:
+    return List(Object.keys(action.voteProposals).map(key => action.voteProposals[key].id || null));
+  case CREATE_VOTE_PROPOSAL:
+    return state.push(action.id);
+  case MOVE_PROPOSAL_UP: {
+    const idx = state.indexOf(action.id);
+    return state.delete(idx).insert(idx - 1, action.id);
+  }
+  case MOVE_PROPOSAL_DOWN: {
+    const idx = state.indexOf(action.id);
+    return state.delete(idx).insert(idx + 1, action.id);
+  }
+  default:
+    return state;
+  }
+};
+
+const defaultVoteProposal = Map({
+  isNew: true,
+  toDelete: false,
+  id: '',
+  titleEntries: List(),
+  descriptionEntries: List(),
+  modules: List()
+});
+
+export const voteProposalsById = (state: Map<string, Map> = Map(), action: ReduxAction<Action>) => {
+  switch (action.type) {
+  case UPDATE_VOTE_PROPOSALS: {
+    let newState = Map();
+    action.voteProposals.forEach((proposal) => {
+      const proposalInfo = fromJS({
+        isNew: false,
+        order: proposal.order,
+        toDelete: false,
+        id: proposal.id,
+        titleEntries: proposal.titleEntries,
+        descriptionEntries: proposal.descriptionEntries,
+        modules: proposal.modules ? proposal.modules.map(m => m.id) : []
+      });
+      newState = newState.set(proposal.id, proposalInfo);
+    });
+    return newState;
+  }
+  case CREATE_VOTE_PROPOSAL:
+    return state.set(action.id, defaultVoteProposal.set('id', action.id));
+  case DELETE_VOTE_PROPOSAL:
+    return state.setIn([action.id, 'toDelete'], true);
+  case UPDATE_VOTE_PROPOSAL_TITLE:
+    return state.updateIn([action.id, 'titleEntries'], updateInLangstringEntries(action.locale, action.value));
+  case UPDATE_VOTE_PROPOSAL_DESCRIPTION:
+    return state.updateIn([action.id, 'descriptionEntries'], updateInLangstringEntries(action.locale, fromJS(action.value)));
```

Comment:
This one is immutable, however `titleEntries` is not? How come?

Reply:
I think it's because titleEntries is not used on proposals. I will remove all this dead code

Reply:
oh no ! it's because description is a rich text (and so its value is an object (raw content state from draft)) and title is just a simple input (simple string)